### ISSUE TITLE
As a developer, convert our namespace handling to not assume npsp, but use whatever packaging org namespace we are in.

### ISSUE THREAD

**Issue by mrbelvedere** [SalesforceFoundation/Cumulus#1310]
Here are the code snippets I mentioned for dynamically looking up namespace:

Our initial Apex code implementation of this revolves around this utility class.

```apex
public with sharing abstract class ApplicationUtilities 
{
    public static String getNamespace()
    {   	
        if (plainNamespace == null)
        {
            String withDotNotation = ApplicationUtilities.class.getName();  
        
            if (withDotNotation.contains('.'))
            {
                plainNamespace = withDotNotation.substringBefore('.');
            }
            else
            {
                plainNamespace = '';
            }
        }
        return plainNamespace;   
    }
    
    public static String getNameSpacePrefix(); // snipped -- returns with  '__' or '' -- similar to above
    public static String getNamespaceDotNotation(); // snipped -- returns with '.' or ''  -- similar to above
}
```
Then anywhere with a Dynamic Query looks up field values through:
```
account.get(ApplicationUtilities.getNameSpacePrefix()+'CustomField__c');
```
or if you need a Class Name through a string:
```
String className = ApplicationUtilities.getNamespaceDotNotation()+'PolicyController';
```
Then one trickier place is OnClick JavaScript buttons. This of course would fail with multiple ApplicationUtilities classes. 

```javascript
{!REQUIRESCRIPT("/soap/ajax/30.0/connection.js")}
{!REQUIRESCRIPT("/soap/ajax/30.0/apex.js")}
var namespace_prefix = ''; 
var gettingnamespace = sforce.connection.query("SELECT NamespacePrefix FROM ApexClass where Name = 'ApplicationUtilities' LIMIT 1"); 
var getname = gettingnamespace.getArray("records"); 

if(getname.length > 0) 
{ 
        namespace_prefix = getname[0].NamespacePrefix; 
}

if (namespace_prefix==null || namespace_prefix=='')
	window.location = 'apex/InsuranceSelectionUI?accountId=' + '{!Account.Id}';
else
	window.location = 'apex/'+namespace_prefix+'__InsuranceSelectionUI?accountId=' + '{!Account.Id}';
```
In any Visualforce page / component we just always pass the namespace from the controller and then use '{!namespacePrefix}' where necessary. Which works for visualforce and JavaScript in the page.

Controller:
```apex
public String namespacePrefix 
{
	get
	{
		return ApplicationUtilities.getNameSpacePrefix();
	} 
}
```
Page:
```
$scope.namespacePrefix = '{!namespacePrefix}';
```
I think those are the really common scenarios, and all seem to work very well for us. As I said before, Flows are the one place where we are actively looking into potential solutions. Let me know if I missed any others. 

Thanks again!

Adam Rutland
Vlocity

**Comment by force2b** [comment]
Very creative solution for the javascript button. One question, will the query on ApexClass always work for any user regardless of their permissions?

**Comment by mrbelvedere** [comment]
Included in beta release version 3.35 (Beta 9) and higher

**Comment by mrbelvedere** [comment]
Included in production release version 3.35
